<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
    
    <title>The Official Radare Blog  | Radare2 and bioinformatics: a good match?</title>
    <meta name="viewport" content="width=device-width,minimum-scale=1">
    <meta name="generator" content="Hugo 0.58.3" />
    
    
      <META NAME="ROBOTS" CONTENT="NOINDEX, NOFOLLOW">
    

    
    
      <link href="/dist/css/app.d98f2eb6bcd1eaedb7edf166bd16af26.css" rel="stylesheet">
    

    

    
      
    

    

    <meta property="og:title" content="Radare2 and bioinformatics: a good match?" />
<meta property="og:description" content="Intro Ahead of this years&rsquo; radarecon, pancake nudged me into discussion we both have about how software reverse engineering and bioinformatics compare and might complement each other, if at all. Inspired by Bunnie Huang&rsquo;s writeups on (computational) biology as a living example of a cross-domain polymath, I&rsquo;ll attempt to write down some thoughts and pointers on how radare could be used (or not) in bioinformatics and hopefully manage expectations on what&rsquo;s possible today." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://radare.today/posts/radare2-bioinformatics/" />
<meta property="article:published_time" content="2018-08-31T00:00:00+00:00" />
<meta property="article:modified_time" content="2018-08-31T00:00:00+00:00" />
<meta itemprop="name" content="Radare2 and bioinformatics: a good match?">
<meta itemprop="description" content="Intro Ahead of this years&rsquo; radarecon, pancake nudged me into discussion we both have about how software reverse engineering and bioinformatics compare and might complement each other, if at all. Inspired by Bunnie Huang&rsquo;s writeups on (computational) biology as a living example of a cross-domain polymath, I&rsquo;ll attempt to write down some thoughts and pointers on how radare could be used (or not) in bioinformatics and hopefully manage expectations on what&rsquo;s possible today.">


<meta itemprop="datePublished" content="2018-08-31T00:00:00&#43;00:00" />
<meta itemprop="dateModified" content="2018-08-31T00:00:00&#43;00:00" />
<meta itemprop="wordCount" content="832">



<meta itemprop="keywords" content="" />
<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Radare2 and bioinformatics: a good match?"/>
<meta name="twitter:description" content="Intro Ahead of this years&rsquo; radarecon, pancake nudged me into discussion we both have about how software reverse engineering and bioinformatics compare and might complement each other, if at all. Inspired by Bunnie Huang&rsquo;s writeups on (computational) biology as a living example of a cross-domain polymath, I&rsquo;ll attempt to write down some thoughts and pointers on how radare could be used (or not) in bioinformatics and hopefully manage expectations on what&rsquo;s possible today."/>

  </head>

  <body class="ma0 avenir bg-near-white">

    
   
  

  <header>
    <div class="bg-black">
      <nav class="pv3 ph3 ph4-ns" role="navigation">
  <div class="flex-l justify-between items-center center">
    <a href="https://radare.today" class="f3 fw2 hover-white no-underline white-90 dib">
      The Official Radare Blog
    </a>
    <div class="flex-l items-center">
      

      
      



<a href="radareorg" target="_blank" class="link-transition twitter link dib z-999 pt3 pt0-l mr1" title="Twitter link" rel="noopener" aria-label="follow on Twitter——Opens in a new window">
  <svg height="32px"  style="enable-background:new 0 0 67 67;" version="1.1" viewBox="0 0 67 67" width="32px" xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><path d="M37.167,22.283c-2.619,0.953-4.274,3.411-4.086,6.101  l0.063,1.038l-1.048-0.127c-3.813-0.487-7.145-2.139-9.974-4.915l-1.383-1.377l-0.356,1.017c-0.754,2.267-0.272,4.661,1.299,6.271  c0.838,0.89,0.649,1.017-0.796,0.487c-0.503-0.169-0.943-0.296-0.985-0.233c-0.146,0.149,0.356,2.076,0.754,2.839  c0.545,1.06,1.655,2.097,2.871,2.712l1.027,0.487l-1.215,0.021c-1.173,0-1.215,0.021-1.089,0.467  c0.419,1.377,2.074,2.839,3.918,3.475l1.299,0.444l-1.131,0.678c-1.676,0.976-3.646,1.526-5.616,1.568  C19.775,43.256,19,43.341,19,43.405c0,0.211,2.557,1.397,4.044,1.864c4.463,1.377,9.765,0.783,13.746-1.568  c2.829-1.673,5.657-5,6.978-8.221c0.713-1.716,1.425-4.851,1.425-6.354c0-0.975,0.063-1.102,1.236-2.267  c0.692-0.678,1.341-1.419,1.467-1.631c0.21-0.403,0.188-0.403-0.88-0.043c-1.781,0.636-2.033,0.551-1.152-0.402  c0.649-0.678,1.425-1.907,1.425-2.267c0-0.063-0.314,0.042-0.671,0.233c-0.377,0.212-1.215,0.53-1.844,0.72l-1.131,0.361l-1.027-0.7  c-0.566-0.381-1.361-0.805-1.781-0.932C39.766,21.902,38.131,21.944,37.167,22.283z M33,64C16.432,64,3,50.569,3,34S16.432,4,33,4  s30,13.431,30,30S49.568,64,33,64z" style="fill-rule:evenodd;clip-rule:evenodd;fill:;"/></svg>

<span class="new-window"><svg  height="8px"  style="enable-background:new 0 0 1000 1000;" version="1.1" viewBox="0 0 1000 1000" width="8px" xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" >
<path d="M598 128h298v298h-86v-152l-418 418-60-60 418-418h-152v-86zM810 810v-298h86v298c0 46-40 86-86 86h-596c-48 0-86-40-86-86v-596c0-46 38-86 86-86h298v86h-298v596h596z" style="fill-rule:evenodd;clip-rule:evenodd;fill:;"/>
</svg>
</span></a>









    </div>
  </div>
</nav>

    </div>
  </header>



    <main class="pb7" role="main">
      
  
  <article class="flex-l flex-wrap justify-between mw8 center ph3">

    <header class="mt4 w-100">
      <p class="f6 b helvetica tracked">
          
        POSTS
      </p>
      <h1 class="f1 athelas mb1">Radare2 and bioinformatics: a good match?</h1>
      
      <time class="f6 mv4 dib tracked" datetime="2018-08-31T00:00:00Z">August 31, 2018</time>
      
      
    </header>

    <section class="nested-copy-line-height lh-copy serif f4 nested-links nested-img mid-gray pr4-l w-two-thirds-l">

<h2 id="intro">Intro</h2>

<p>Ahead of this <a href="https://rada.re/con/2018/">years&rsquo; radarecon</a>, pancake nudged me into discussion we both have about how <a href="https://www.reddit.com/r/reverseengineering">software reverse engineering</a> and bioinformatics compare and might complement each other, if at all. Inspired by <a href="https://www.bunniestudios.com/blog/?cat=16">Bunnie Huang&rsquo;s writeups</a> on (computational) biology as a living example of a cross-domain polymath, I&rsquo;ll attempt to write down some thoughts and <strong>pointers on how radare could be used (or not) in bioinformatics</strong> and hopefully manage expectations on what&rsquo;s possible today.</p>

<p>For starters, back in 2015, the simple <a href="https://www.illumina.com/informatics/sequencing-data-analysis/sequence-file-formats.html">Illumina BCL file format</a> got included in <a href="https://github.com/radare/radare2-extras">radare-extras</a>. As I was providing some specs and explaning <a href="https://en.wikipedia.org/wiki/DNA_sequencing">how DNA sequencing worked</a> in general, <a href="https://github.com/radare/radare2-extras/blob/master/bcl/core_bcl.c">pancake quickly put together a radare plugin</a> for this fairly straightforward file format.</p>

<p><img src="/images/bcl_format_tldr_csr4.jpg" alt="BCL format" /></p>

<p>Then fast forward into 2018, <a href="https://github.com/radare/radare2-extras/issues/165">radare seems to want more</a>. Here comes the crux of the matter:</p>

<blockquote>
<p><strong>Since DNA and ASM are just code running on different architectures, we can reuse the same radare2 reversing workflow right? RIGHT?</strong></p>
</blockquote>

<p>Are they?</p>

<p>Before jumping into sleepless nights of unstoppable implementation, please take the time to read two fun papers touching both domains of reverse engineering and biology. They <strong>reveal how different (or similar) electronics manufacturing and biology can be</strong>.</p>

<p>As stated in <a href="https://www.cell.com/cancer-cell/abstract/S1535-6108(02)00133-2?code=cell-site">biologists trying to figure out a radio</a> paper:</p>

<blockquote>
<p>&rdquo;(&hellip;) the commonality of the language allows engineers to identify
familiar patterns or modules (a trigger, an amplifier, etc.) in a
diagram of an unfamiliar device&rdquo;</p>
</blockquote>

<p>Food for thought: <strong>how &ldquo;unfamiliar&rdquo; of a &ldquo;device&rdquo; is biology itself</strong> versus human-<strong>manufactured</strong> consumer wares?</p>

<p>In another, more recent, neuroscinece paper, <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005268">&ldquo;Could a Neuroscientist Understand a Microprocessor?&rdquo;</a>, some insights come up:</p>

<blockquote>
<p>Much has been written about the differences between computation in silico and computation in vivo (&hellip;) the stochasticity, redundancy, and robustness present in biological systems seems dramatically different from that of a microprocessor. But there are many parallels we can draw between the two types of systems.</p>
</blockquote>

<p>Bottom line is, they are definitely different yet similar in <strong>some</strong> instances. Without getting overwhelmed by the huge, sometimes messy, amount of domain-specific knowledge to digest, <strong>how can we score &ldquo;quick wins&rdquo; for radare2</strong> if an implementation under radare-extras starts to happen?</p>

<p>Observing bioinformatics <a href="https://rse.ac.uk/">from the RSE perspective</a>, I see great contributions that could happen in three areas:</p>

<ol>
<li>EBA: Exploratory Bioinformatics Analysis.</li>
<li>Scientific algorithm optimization and security.</li>
<li>Outreach.</li>
</ol>

<h2 id="radare2-tl-dr-for-bioinformaticians">Radare2 TL;DR for bioinformaticians</h2>

<p>Here&rsquo;s some explanation on <a href="https://radare.gitbooks.io/radare2book/content/">how radare works from a user perspective</a>. If you barely recall what assembly was from school, <a href="https://twitter.com/azeria_labs">I&rsquo;ll leave you in good hands</a> to catch up with <a href="https://azeria-labs.com/writing-arm-assembly-part-1/">ARM assembly here</a></p>

<p>Also, some relatively recent UI eyecandy from <a href="https://github.com/radareorg/cutter">Cutter</a>:</p>

<p><img src="https://raw.githubusercontent.com/radareorg/cutter/master/docs/screenshot.png" alt="Cutter radare2" /></p>

<p>Often touted as &ldquo;steep learning curve&rdquo; framework due to its commands, radare2 has been misunderstood for years, since in reality, keybindings allow for distraction-free <strong>fast iteration</strong> during binary analysis.</p>

<h2 id="bioinfomatics-tl-dr-for-radare2-developers">Bioinfomatics TL;DR for radare2 developers</h2>

<p>If you are a r2 developer, those are the formats radare would need to understand and implement to be minimally interesting for our biologist neighbors (optional ones, inside parenthesis):</p>

<blockquote>
<p><a href="[https://samtools.github.io/hts-specs/SAMv1.pdf]">SAM and BAM</a>, (<a href="https://bioinformatics-workbook.readthedocs.io/en/latest/introduction/fileFormats/">FASTA, FASTQ, VCF</a>, <a href="http://www.internationalgenome.org/faq/what-are-cram-files">CRAM</a>, <a href="https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/bty608/5051198">Crumble</a>)</p>
</blockquote>

<p>Now, one could go <strong>the hardcore pancake/Feynman <a href="https://github.com/brentp">(brentp?)</a> way</strong> and implement file parsers from scratch <strong>or</strong> use some third party library such as <a href="http://www.htslib.org/">htslib</a>.</p>

<p>After basic read/write functionality is in place, I think that a potential first win would be to have the <a href="https://en.wikipedia.org/wiki/Midnight_Commander">&ldquo;Midnight Commander&rdquo;</a>-equivalent of radare4bio for curious and impatient bioinformaticians.</p>

<p>There&rsquo;s great educational potential if this is implemented right since <strong>radare allows for fast VIM-like iteration and speed during complex analysis</strong>.</p>

<p>For instance, being able to <strong>examine individual reads with VIM shortcuts</strong>, flip/cycle <a href="https://wiki.bits.vib.be/index.php/CIGAR">CIGAR encodings</a>, like with the radare2 bit editor:</p>

<p><img src="/images/r2_bit_editor.png" alt="radare2 bit editor" /></p>

<p>Group reads by some arbitrary criteria, subsample, filter them, write out, etc&hellip;</p>

<p><strong>That is, FAST exploratory bioinformatics analysis (EBA) without the overhead of writing discrete commands or putting together <a href="https://www.commonwl.org/">workflows</a>, <a href="https://github.com/bcbio/bcbio-nextgen">pipelines</a> and/or lengthy documentation</strong></p>

<h2 id="outro">Outro</h2>

<ol>
<li><p>How would radare <strong>really help</strong> with &ldquo;biology reverse engineering&rdquo;?</p></li>

<li><p>How can radare absorb those &ldquo;extras&rdquo; <a href="https://bioconda.github.io/">without introducing a vast dependency tree of bioinfo software</a>? Perhaps a <strong>clean-room implementation</strong> is still of interest nowadays?</p></li>

<li><p>Would all that coding effort be worth it?</p></li>
</ol>

<p>Those are open questions at the time of writing this, but here are some opportunities:</p>

<p>When bioinformaticists analyze data (and are not waiting for big computations to complete), <strong>it is hugely helpful to iterate fast on a particular question</strong>. Keeping focus on the task at hand while answering questions fast and accurately is immensely valuable.</p>

<p>Radare2 is well positioned in this regard, allowing for fast adhoc analysis for the reasons stated before (VIM-like blazing speed shortcuts, focus on speed).</p>

<p>As <a href="http://mussol.org/2016/06/11/changing-career-paths/">a former colleague pointed out</a>, Bioformatics (<a href="https://www.nature.com/news/2010/101013/full/467775a.html">and scientific software in general</a>) is in dire <strong>need for optimization and good software engineering at several levels</strong>: Storage, data processing, security (<a href="https://dnasec.cs.washington.edu/dnasec.pdf">read this!</a>), good software design patterns, etc&hellip;</p>

<p><strong>If all else fails, the outreach value of <a href="https://www.nature.com/articles/d41586-018-05769-8">getting reverse engineers poking into computational biology</a> is in itself, a huge win, IMHO.</strong></p>

<p>If you are still reading this, I&rsquo;m honored :)</p>

<p>Please ping me physically <a href="https://rada.re/con/2018/">during the radare2 2018 con</a> or via <a href="http://twitter.com/braincode">twitter <code>@braincode</code></a> if you want to have a chat about this and other random braindumplings.</p>
<ul class="pa0">
  
</ul>
<div class="mt6">
      
      
      </div>
    </section>

    <aside class="w-30-l mt6-l">




</aside>

  </article>

    </main>
    <footer class="bg-black bottom-0 w-100 pa3" role="contentinfo">
  <div class="flex justify-between">
  <a class="f4 fw4 hover-white no-underline white-70 dn dib-ns pv2 ph3" href="https://radare.today" >
    &copy; 2019 The Official Radare Blog
  </a>
    <div>



<a href="radareorg" target="_blank" class="link-transition twitter link dib z-999 pt3 pt0-l mr1" title="Twitter link" rel="noopener" aria-label="follow on Twitter——Opens in a new window">
  <svg height="32px"  style="enable-background:new 0 0 67 67;" version="1.1" viewBox="0 0 67 67" width="32px" xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><path d="M37.167,22.283c-2.619,0.953-4.274,3.411-4.086,6.101  l0.063,1.038l-1.048-0.127c-3.813-0.487-7.145-2.139-9.974-4.915l-1.383-1.377l-0.356,1.017c-0.754,2.267-0.272,4.661,1.299,6.271  c0.838,0.89,0.649,1.017-0.796,0.487c-0.503-0.169-0.943-0.296-0.985-0.233c-0.146,0.149,0.356,2.076,0.754,2.839  c0.545,1.06,1.655,2.097,2.871,2.712l1.027,0.487l-1.215,0.021c-1.173,0-1.215,0.021-1.089,0.467  c0.419,1.377,2.074,2.839,3.918,3.475l1.299,0.444l-1.131,0.678c-1.676,0.976-3.646,1.526-5.616,1.568  C19.775,43.256,19,43.341,19,43.405c0,0.211,2.557,1.397,4.044,1.864c4.463,1.377,9.765,0.783,13.746-1.568  c2.829-1.673,5.657-5,6.978-8.221c0.713-1.716,1.425-4.851,1.425-6.354c0-0.975,0.063-1.102,1.236-2.267  c0.692-0.678,1.341-1.419,1.467-1.631c0.21-0.403,0.188-0.403-0.88-0.043c-1.781,0.636-2.033,0.551-1.152-0.402  c0.649-0.678,1.425-1.907,1.425-2.267c0-0.063-0.314,0.042-0.671,0.233c-0.377,0.212-1.215,0.53-1.844,0.72l-1.131,0.361l-1.027-0.7  c-0.566-0.381-1.361-0.805-1.781-0.932C39.766,21.902,38.131,21.944,37.167,22.283z M33,64C16.432,64,3,50.569,3,34S16.432,4,33,4  s30,13.431,30,30S49.568,64,33,64z" style="fill-rule:evenodd;clip-rule:evenodd;fill:;"/></svg>

<span class="new-window"><svg  height="8px"  style="enable-background:new 0 0 1000 1000;" version="1.1" viewBox="0 0 1000 1000" width="8px" xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" >
<path d="M598 128h298v298h-86v-152l-418 418-60-60 418-418h-152v-86zM810 810v-298h86v298c0 46-40 86-86 86h-596c-48 0-86-40-86-86v-596c0-46 38-86 86-86h298v86h-298v596h596z" style="fill-rule:evenodd;clip-rule:evenodd;fill:;"/>
</svg>
</span></a>








</div>
  </div>
</footer>

    

  <script src="/dist/js/app.3fc0f988d21662902933.js"></script>


  </body>
</html>
